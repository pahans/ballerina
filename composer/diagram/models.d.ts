// This is an autogenerated file. Do not edit. Run 'npm run gen-models' to generate.
    export as namespace Ballerina;
    export interface CompilationUnit {
    ws: (object)[]
    position: object
    id: string
    name: string
    topLevelNodes: (Import|Endpoint|Service|Variable|Function|TypeDefinition)[]
    kind: string
}
export interface Import {
    ws: (object)[]
    position: object
    id: string
    orgName: Identifier
    alias: Identifier
    packageName: (Identifier)[]
    kind: string
    packageVersion: Identifier
}
export interface Identifier {
    id: string
    literal: boolean
    value: string
    kind: string
    valueWithBar: string
    position?: object
}
export interface TypeDefinition {
    ws: (object)[]
    position: object
    id: string
    name: Identifier
    deprecatedAttachments: any
    annotationAttachments: any
    typeNode: ObjectType
    kind: string
    public: boolean
    private: boolean
    native: boolean
    final: boolean
    attached: boolean
    lambda: boolean
    parallel: boolean
    connector: boolean
    deprecated: boolean
    readonly: boolean
    function_final: boolean
    interface: boolean
    defaultable_checked: boolean
    defaultable: boolean
    record: boolean
    compensate: boolean
    abstract: boolean
    optional: boolean
    testable: boolean
    isObjectType: boolean
    isAbstractKeywordAvailable?: boolean
}
export interface ObjectType {
    position: object
    id: string
    symbolType: (string)[]
    functions: (Function)[]
    fields: (Variable)[]
    grouped: boolean
    nullable: boolean
    kind: string
    typeReferences: any
    initFunction?: Function
}
export interface Function {
    ws?: (object)[]
    position: object
    id: string
    workers: (Worker)[]
    returnTypeAnnotationAttachments: any
    name: Identifier
    annotationAttachments: (AnnotationAttachment)[]
    returnTypeNode: ValueType
    receiver?: Variable
    defaultableParameters: any
    deprecatedAttachments: any
    endpointNodes: any
    parameters: (Variable)[]
    kind: string
    public: boolean
    private: boolean
    native: boolean
    final: boolean
    attached: boolean
    lambda: boolean
    parallel: boolean
    connector: boolean
    deprecated: boolean
    readonly: boolean
    function_final: boolean
    interface: boolean
    defaultable_checked: boolean
    defaultable: boolean
    record: boolean
    compensate: boolean
    abstract: boolean
    optional: boolean
    testable: boolean
    hasReturns?: boolean
    allParams: (Variable)[]
    noVisibleReceiver?: boolean
    body?: Block
    isConstructor?: boolean
    objectOuterFunction?: boolean
    objectOuterFunctionTypeName?: Identifier
    restParameters?: Variable
}
export interface ValueType {
    ws?: (object)[]
    position?: object
    id: string
    symbolType?: (string)[]
    typeKind: string
    grouped: boolean
    nullable: boolean
    kind: string
    withParantheses?: boolean
}
export interface Variable {
    position: object
    id: string
    symbolType?: (string)[]
    safeAssignment: boolean
    deprecatedAttachments: any
    annotationAttachments: any
    name: Identifier
    kind: string
    public: boolean
    private: boolean
    native: boolean
    final: boolean
    attached: boolean
    lambda: boolean
    parallel: boolean
    connector: boolean
    deprecated: boolean
    readonly: boolean
    function_final: boolean
    interface: boolean
    defaultable_checked: boolean
    defaultable: boolean
    record: boolean
    compensate: boolean
    abstract: boolean
    optional: boolean
    testable: boolean
    ws?: (object)[]
    typeNode?: ValueType|UserDefinedType|FunctionType|ArrayType|ConstrainedType|BuiltInRefType
    inObject?: boolean
    endWithSemicolon?: boolean
    noVisibleName?: boolean
    initialExpression?: TypeInitExpr|Lambda|ArrowExpr|ArrayLiteralExpr|Literal|CheckExpr|SimpleVariableRef|Invocation|AwaitExpr|RecordLiteralExpr
    arrowExprParam?: boolean
    global?: boolean
    rest?: boolean
    serviceEndpoint?: boolean
}
export interface Block {
    id: string
    statements: (VariableDef|ExpressionStatement|Assignment|CompoundAssignment|While|WorkerReceive|Return|WorkerSend|Match|If)[]
    kind: string
    position?: object
}
export interface Return {
    ws: (object)[]
    position: object
    id: string
    expression: BinaryExpr|SimpleVariableRef
    kind: string
}
export interface BinaryExpr {
    ws: (object)[]
    position: object
    id: string
    symbolType: (string)[]
    rightExpression: SimpleVariableRef|Literal|Invocation
    operatorKind: string
    kind: string
    leftExpression: BinaryExpr|SimpleVariableRef|FieldBasedAccessExpr|Literal
}
export interface SimpleVariableRef {
    ws: (object)[]
    position: object
    id: string
    symbolType?: (string)[]
    variableName: Identifier
    packageAlias: Identifier
    kind: string
    isExpression?: boolean
}
export interface Literal {
    ws: (object)[]
    position: object
    id: string
    symbolType: (string)[]
    value: string
    unescapedValue?: string
    originalValue: string
    kind: string
    isExpression?: boolean
}
export interface UserDefinedType {
    ws: (object)[]
    position: object
    id: string
    symbolType: (string)[]
    packageAlias: Identifier
    grouped: boolean
    public: boolean
    private: boolean
    native: boolean
    final: boolean
    attached: boolean
    lambda: boolean
    parallel: boolean
    connector: boolean
    deprecated: boolean
    readonly: boolean
    function_final: boolean
    interface: boolean
    defaultable_checked: boolean
    defaultable: boolean
    record: boolean
    compensate: boolean
    abstract: boolean
    optional: boolean
    testable: boolean
    typeName: Identifier
    nullable: boolean
    kind: string
}
export interface If {
    ws: (object)[]
    position: object
    id: string
    body: Block
    condition: BracedTupleExpr
    kind: string
}
export interface Assignment {
    ws: (object)[]
    position: object
    id: string
    variable: FieldBasedAccessExpr|IndexBasedAccessExpr|SimpleVariableRef
    expression: SimpleVariableRef|IndexBasedAccessExpr|Literal|ArrayLiteralExpr|AwaitExpr
    declaredWithVar: boolean
    kind: string
}
export interface FieldBasedAccessExpr {
    ws: (object)[]
    position: object
    id: string
    symbolType: (string)[]
    fieldName: Identifier
    expression: SimpleVariableRef
    kind: string
}
export interface BracedTupleExpr {
    ws: (object)[]
    position: object
    id: string
    symbolType: (string)[]
    expressions: (Literal|BinaryExpr)[]
    kind: string
}
export interface VariableDef {
    ws: (object)[]
    position: object
    id: string
    variable: Variable
    kind: string
}
export interface TypeInitExpr {
    ws: (object)[]
    position: object
    id: string
    symbolType: (string)[]
    expressions: (Literal)[]
    type: UserDefinedType
    kind: string
    hasParantheses: boolean
    typeName: Identifier
    isExpression: boolean
}
export interface ExpressionStatement {
    ws: (object)[]
    position: object
    id: string
    expression: Invocation|MatchExpression
    kind: string
}
export interface Invocation {
    ws: (object)[]
    position: object
    id: string
    symbolType: any
    invocationType?: string
    iterableOperation: boolean
    argumentExpressions: (Literal|NamedArgsExpr|SimpleVariableRef|IndexBasedAccessExpr|FieldBasedAccessExpr|Invocation|UnaryExpr|BinaryExpr)[]
    async: boolean
    packageAlias: Identifier
    actionInvocation: boolean
    name: Identifier
    expression?: SimpleVariableRef
    kind: string
    isExpression?: boolean
}
export interface FunctionType {
    ws: (object)[]
    position: object
    id: string
    symbolType: (string)[]
    returnKeywordExists: boolean
    params: (Variable)[]
    grouped: boolean
    returnTypeNode: ValueType
    nullable: boolean
    kind: string
    hasReturn: boolean
}
export interface Lambda {
    position: object
    id: string
    symbolType: (string)[]
    functionNode: Function
    kind: string
    isExpression: boolean
}
export interface ArrowExpr {
    ws: (object)[]
    position: object
    id: string
    symbolType: (string)[]
    parameters: (Variable)[]
    expression: BinaryExpr
    kind: string
    hasParantheses: boolean
    isExpression: boolean
}
export interface ArrayType {
    ws?: (object)[]
    position?: object
    id: string
    symbolType: (string)[]
    dimensions: number
    elementType: ValueType
    grouped: boolean
    sizes: (number)[]
    nullable: boolean
    kind: string
    dimensionAsString?: string
    isRestParam?: boolean
}
export interface ArrayLiteralExpr {
    ws: (object)[]
    position: object
    id: string
    symbolType: (string)[]
    expressions: (Literal|ArrayLiteralExpr)[]
    kind: string
    isExpression?: boolean
}
export interface AnnotationAttachment {
    ws: (object)[]
    position: object
    id: string
    expression?: RecordLiteralExpr
    annotationName: Identifier
    packageAlias: Identifier
    kind: string
}
export interface RecordLiteralExpr {
    ws: (object)[]
    position: object
    id: string
    symbolType: (string)[]
    keyValuePairs: (RecordLiteralKeyValue)[]
    kind: string
    isExpression?: boolean
}
export interface RecordLiteralKeyValue {
    ws: (object)[]
    id: string
    value: Literal
    key: SimpleVariableRef|Literal
    kind: string
}
export interface IndexBasedAccessExpr {
    ws: (object)[]
    position: object
    id: string
    symbolType: (string)[]
    expression: SimpleVariableRef|IndexBasedAccessExpr
    index: SimpleVariableRef|Literal
    kind: string
}
export interface CompoundAssignment {
    ws: (object)[]
    position: object
    id: string
    expression: Literal
    operatorKind: string
    variable: SimpleVariableRef
    kind: string
    compoundOperator: string
}
export interface CheckExpr {
    ws: (object)[]
    position: object
    id: string
    symbolType: (string)[]
    operatorKind: string
    kind: string
    expression: TypeConversionExpr
    isExpression: boolean
}
export interface TypeConversionExpr {
    ws: (object)[]
    position: object
    id: string
    symbolType: (string)[]
    expression: SimpleVariableRef|IndexBasedAccessExpr
    typeNode: ValueType|BuiltInRefType
    kind: string
}
export interface UnaryExpr {
    ws: (object)[]
    position: object
    id: string
    symbolType: (string)[]
    operatorKind: string
    kind: string
    expression: SimpleVariableRef|IndexBasedAccessExpr|Invocation
}
export interface Endpoint {
    ws: (object)[]
    position: object
    id: string
    symbolType: (string)[]
    configurationExpression: RecordLiteralExpr
    name: Identifier
    annotationAttachments: any
    kind: string
    endPointType: UserDefinedType
    public: boolean
    private: boolean
    native: boolean
    final: boolean
    attached: boolean
    lambda: boolean
    parallel: boolean
    connector: boolean
    deprecated: boolean
    readonly: boolean
    function_final: boolean
    interface: boolean
    defaultable_checked: boolean
    defaultable: boolean
    record: boolean
    compensate: boolean
    abstract: boolean
    optional: boolean
    testable: boolean
}
export interface ConstrainedType {
    ws: (object)[]
    position: object
    id: string
    symbolType: (string)[]
    grouped: boolean
    nullable: boolean
    constraint: ValueType|UnionTypeNode
    type: BuiltInRefType
    kind: string
}
export interface BuiltInRefType {
    ws: (object)[]
    position: object
    id: string
    symbolType: (string)[]
    grouped: boolean
    typeKind: string
    nullable: boolean
    kind: string
}
export interface UnionTypeNode {
    ws: (object)[]
    position: object
    id: string
    symbolType: (string)[]
    memberTypeNodes: (UserDefinedType)[]
    grouped: boolean
    nullable: boolean
    kind: string
}
export interface AwaitExpr {
    ws: (object)[]
    position: object
    id: string
    symbolType: (string)[]
    expression: SimpleVariableRef
    kind: string
    isExpression?: boolean
}
export interface Match {
    ws: (object)[]
    position: object
    id: string
    expression: SimpleVariableRef
    kind: string
    patternClauses: (MatchPatternClause)[]
}
export interface MatchPatternClause {
    ws: (object)[]
    position: object
    id: string
    variableNode: Variable
    statement: Block
    kind: string
    withCurlies: boolean
}
export interface Worker {
    ws: (object)[]
    position: object
    id: string
    workers: any
    returnTypeAnnotationAttachments: any
    name: Identifier
    annotationAttachments: any
    returnTypeNode: ValueType
    defaultableParameters: any
    deprecatedAttachments: any
    endpointNodes: any
    body: Block
    parameters: (Variable)[]
    kind: string
    public: boolean
    private: boolean
    native: boolean
    final: boolean
    attached: boolean
    lambda: boolean
    parallel: boolean
    connector: boolean
    deprecated: boolean
    readonly: boolean
    function_final: boolean
    interface: boolean
    defaultable_checked: boolean
    defaultable: boolean
    record: boolean
    compensate: boolean
    abstract: boolean
    optional: boolean
    testable: boolean
}
export interface WorkerSend {
    ws: (object)[]
    position: object
    id: string
    workerName: Identifier
    expression: SimpleVariableRef
    forkJoinedSend: boolean
    kind: string
}
export interface WorkerReceive {
    ws: (object)[]
    position: object
    id: string
    expression: SimpleVariableRef
    workerName: Identifier
    kind: string
}
export interface While {
    ws: (object)[]
    position: object
    id: string
    condition: BracedTupleExpr
    kind: string
    body: Block
}
export interface Service {
    ws: (object)[]
    position: object
    id: string
    annotationAttachments: any
    initFunction: Function
    endpointNodes: any
    namespaceDeclarations: any
    boundEndpoints: (SimpleVariableRef)[]
    resources: (Resource)[]
    deprecatedAttachments: any
    variables: any
    name: Identifier
    kind: string
    isServiceTypeUnavailable: boolean
}
export interface Resource {
    ws: (object)[]
    position: object
    id: string
    workers: any
    returnTypeAnnotationAttachments: any
    name: Identifier
    annotationAttachments: any
    returnTypeNode: ValueType
    defaultableParameters: any
    deprecatedAttachments: any
    endpointNodes: any
    body: Block
    parameters: (Variable)[]
    kind: string
    public: boolean
    private: boolean
    native: boolean
    final: boolean
    attached: boolean
    lambda: boolean
    parallel: boolean
    connector: boolean
    deprecated: boolean
    readonly: boolean
    function_final: boolean
    interface: boolean
    defaultable_checked: boolean
    defaultable: boolean
    record: boolean
    compensate: boolean
    abstract: boolean
    optional: boolean
    testable: boolean
}
export interface MatchExpression {
    ws: (object)[]
    position: object
    id: string
    symbolType: (string)[]
    expression: Invocation
    kind: string
    patternClauses: (MatchExpressionPatternClause)[]
}
export interface MatchExpressionPatternClause {
    ws: (object)[]
    position: object
    id: string
    variableNode: Variable
    kind: string
    statement: Invocation
}
export interface NamedArgsExpr {
    ws: (object)[]
    position: object
    id: string
    symbolType: (string)[]
    expression: SimpleVariableRef
    name: Identifier
    kind: string
}