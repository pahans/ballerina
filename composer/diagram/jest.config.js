module.exports = {
    'roots': [
        '<rootDir>/src-ts',
        '<rootDir>/tests'
    ],
    'transform': {
        '.*\.tsx?$': 'ts-jest'
    },
    'testRegex': '(/__tests__/.*|(\\.|/)(test|spec))\\.(jsx?|tsx?)$',
    'moduleFileExtensions': [
        'ts',
        'tsx',
        'js',
        'jsx',
        'json',
        'node'
    ],
    "collectCoverageFrom": [
        "!**/lib/**",
        "**/src/**",
        "!**/node_modules/**",
        "!**/stories/**"
    ]
}