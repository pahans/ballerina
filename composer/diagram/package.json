{
  "name": "@ballerina/diagram",
  "version": "0.983.1-SNAPSHOT",
  "description": "Provides a set of react components to draw Ballerina Diagrams.",
  "keywords": [
    "ballerina",
    "ballerina-lang",
    "composer"
  ],
  "repository": "ballerina-platform/ballerina-lang",
  "main": "lib/index.js",
  "scripts": {
    "clean": "rimraf lib && rimraf coverage",
    "prebuild": "npm run clean && npm run lint",
    "compile": "tsc -p ./",
    "build": "npm run compile",
    "watch": "tsc -watch -p ./",
    "test": "jest",
    "test-coverage": "npm run test -- --coverage",
    "storybook": "start-storybook -p 9001 -c .storybook",
    "lint": "tslint -t stylish --project .",
    "gen-bbe-asts": "npm run compile && node lib-ts/tools/update-bbe-asts.js"
  },
  "author": "ballerina.io",
  "license": "Apache-2.0",
  "peerDependencies": {
    "react-dom": "^16.2.0",
    "react": "^16.2.0"
  },
  "devDependencies": {
    "@babel/core": "^7.1.5",
    "@storybook/addon-info": "^4.0.4",
    "@storybook/react": "^4.0.4",
    "@types/jest": "^23.3.9",
    "@types/react-test-renderer": "^16.0.3",
    "@types/storybook__react": "^4.0.0",
    "@types/react": "16.4.18",
    "@types/react-dom": "^16.0.9",
    "awesome-typescript-loader": "^5.2.1",
    "babel-loader": "^8.0.4",
    "jest": "^23.6.0",
    "react": "^16.2.0",
    "react-docgen-typescript-webpack-plugin": "^1.1.0",
    "react-dom": "^16.6.1",
    "require-context": "^1.1.0",
    "ts-jest": "^23.10.4",
    "ts-loader": "^5.3.0",
    "tslint": "^5.11.0",
    "typescript": "^3.1.6",
    "vscode-uri": "^1.0.6"
  },
  "dependencies": {
    "@ballerina/ast-model": "file:../ast-model",
    "@ballerina/lang-service": "file:../lang-service",
    "@ballerina/theme": "file:../theme",
    "axios": "^0.16.2",
    "jquery": "^3.3.1",
    "jsplumb": "2.3.5",
    "log4javascript": "^1.4.15",
    "react-autosuggest": "^9.0.1",
    "react-custom-scrollbars": "^4.1.2",
    "react-test-renderer": "^16.6.1",
    "react-transition-group": "^1.1.2",
    "semantic-ui-react": "^0.81.3",
    "uuid": "^3.1.0",
    "winston": "^3.1.0"
  }
}
