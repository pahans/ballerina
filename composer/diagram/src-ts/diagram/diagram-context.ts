import { BallerinaAST } from "@ballerina/ast-model";
import React from "react";

export enum DiagramMode { ACTION, DEFAULT }

export interface IDiagramContext {
    ast: BallerinaAST | undefined;
    changeMode: (newMode: DiagramMode) => void;
    editingEnabled: boolean;
    mode: DiagramMode;
    toggleEditing: () => void;
}

const defaultDiagramContext: IDiagramContext = {
    ast: undefined,
    changeMode: () => {
        // do nothing
    },
    editingEnabled: false,
    mode: DiagramMode.ACTION,
    toggleEditing: () => {
        // do nothing
    },
};

export const DiagramContext = React.createContext(defaultDiagramContext);
