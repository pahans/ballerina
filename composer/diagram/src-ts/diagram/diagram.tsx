import { BallerinaAST, BallerinaASTNode, BallerinaEndpoint, BallerinaSourceFragment } from "@ballerina/ast-model";
import { ASTDidChangeParams, ASTDidChangeResponse, GetASTParams,
    GetASTResponse } from "@ballerina/lang-service";
import React from "react";

export interface DiagramLangClient {
    getAST(params: GetASTParams): Thenable<GetASTResponse>;
    astDidChange(params: ASTDidChangeParams): Thenable<ASTDidChangeResponse>;
    parseFragment(params: BallerinaSourceFragment): Thenable<BallerinaASTNode>;
    getEndpoints(): Thenable<BallerinaEndpoint[]>;
    goToSource(line: number, column: number): void;
}

export interface DiagramProps {
    docUri: string;
    langClient: DiagramLangClient;
    height: number;
    width: number;
}

export interface DiagramState {
    currentAST: BallerinaAST;
    editMode: boolean;
    diagramMode: "action" | "default";
}

export class Diagram extends React.Component<DiagramProps, DiagramState> {
    public render() {
        return <React.Fragment>
            <div>{"Diagram Editor"}</div>
        </React.Fragment>;
    }
}
