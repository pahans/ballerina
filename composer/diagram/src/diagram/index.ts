export * from "./diagram";
export * from "./editable-diagram";
export * from "./diagram-context";
