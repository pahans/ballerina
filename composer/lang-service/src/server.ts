import { ChildProcess, execSync, spawn } from 'child_process';
import * as fs from 'fs';
// tslint:disable-next-line:no-submodule-imports
import { ConnectionCloseHandler, ConnectionErrorHandler, createConnection, IConnection } from 'monaco-languageclient/lib/connection';
import { createMessageConnection, StreamMessageReader, StreamMessageWriter } from 'vscode-jsonrpc';

import * as path from 'path';
import { BallerinaLangClient } from './client';

const LS_DEBUG = process.env.LS_DEBUG === "true";
const LS_CUSTOM_CLASSPATH = process.env.LS_CUSTOM_CLASSPATH;

function createServer(ballerinaHome: string) : ChildProcess {
    let cmd;
    const cwd = path.join(ballerinaHome, 'lib', 'tools', 'lang-server', 'launcher');
    const args: string[] = [];
    if (process.platform === 'win32') {
        cmd = path.join(cwd, 'language-server-launcher.bat');
    } else {
        cmd = 'sh';
        args.push(path.join(cwd, 'language-server-launcher.sh'));
    }

    if (LS_DEBUG) {
        args.push('--debug');
    }
    if (LS_CUSTOM_CLASSPATH) {
        args.push('--classpath', LS_CUSTOM_CLASSPATH);
    }
    return spawn(cmd, args, { cwd });
}

function createLSConnection(ballerinaHome: string) : LangServerProcessConnection {
    const childProcess = createServer(ballerinaHome);
    childProcess.on('error', (err) => {
        // log('Error while starting LS', err);
    });
    const reader = new StreamMessageReader(childProcess.stdout);
    const writer = new StreamMessageWriter(childProcess.stdin);
    const messageConntection = createMessageConnection(reader, writer);
    const errorHandler: ConnectionErrorHandler = (err, msg, count) => {
       // log('Error while starting LS', err);
    };
    const closeHandler: ConnectionCloseHandler = () => {
       // log('LS Connection closed');
    };
    const lsConnection = createConnection(messageConntection, errorHandler, closeHandler);
    return {
        childProcess,
        lsConnection
    }
}

interface LangServerProcessConnection {
    childProcess: ChildProcess;
    lsConnection: IConnection;
}


export function detectBallerinaHome(): string {
    // try to ditect the environment.
    const platform: string = process.platform;
    let balHome = '';
    switch (platform) {
        case 'win32': // Windows
            if (process.env.BALLERINA_HOME) {
                return process.env.BALLERINA_HOME;
            }
            try {
                balHome = execSync('where ballerina').toString().trim();
            } catch (error) {
                return balHome;
            }
            if (path) {
                balHome = balHome.replace(/bin\\ballerina.bat$/, '');
            }
            break;
        case 'darwin': // Mac OS
        case 'linux': // Linux
            // lets see where the ballerina command is.
            try {
                const output = execSync('which ballerina');
                balHome = fs.realpathSync(output.toString().trim());
                // remove ballerina bin from path
                if (path) {
                    balHome = balHome.replace(/bin\/ballerina$/, '');
                }
                break;
            } catch {
                return balHome;
            }
    }

    // If we cannot find ballerina home return empty.
    return balHome;
}

export function startBallerinaLangServer(ballerinaHome: string): Thenable<BallerinaLangClient | undefined> {
    const { childProcess, lsConnection } = createLSConnection(ballerinaHome);
    lsConnection.listen();
    lsConnection.onLogMessage((params) => {
        // log('LS LOG: ' +  params.type + ' : ' + params.message);
    });
    return lsConnection.initialize({

        capabilities: {
        },
        processId: process.pid,
        rootUri: null,
        workspaceFolders: null,
    }).then((initializedResult) => {
        if (!initializedResult.capabilities) {
            // log('Error while sending LS init request.');
        }
        return new BallerinaLangClient(childProcess, lsConnection, initializedResult);
    }, (reason) => {
        // log('Error while sending LS init request.');
        // log(reason);
        return undefined;
    });
};