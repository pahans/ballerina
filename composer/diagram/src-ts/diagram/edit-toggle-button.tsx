import React from "react";
import { DiagramContext, IDiagramContext } from "./diagram-context";

export class EditToggleButton extends React.Component {

    private static contextType = DiagramContext;

    public render() {
        const context: IDiagramContext = this.context;
        return <React.Fragment>
            <button onClick={context.toggleEditing} >Toggle Edit Enabled</button>
        </React.Fragment>;
    }
}
