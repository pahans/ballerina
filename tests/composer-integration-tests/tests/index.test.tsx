import * as path from 'path';
import { sync as globSync } from 'glob';
import { startBallerinaLangServer, MinimalLangClient, balToolsPath, BallerinaAST } from './server';
import URI from 'vscode-uri';
import * as React from 'react';
import { BallerinaDiagramWrapper } from '@ballerina/diagram';
import SizingUtil from '@ballerina/diagram/lib/plugins/ballerina/diagram/views/default/sizing-util';
import { create } from 'react-test-renderer';

let langClient : MinimalLangClient;

const bbeDir = path.join(balToolsPath, 'examples');

beforeAll((done) => {
    startBallerinaLangServer()
    .then((languageClient) => {
        if (languageClient) {
            langClient = languageClient;
        } else {
            console.log('Could not start LS properly');
        }
        done();
    }, (reason) => {
        console.log('Could not start LS properly');
        console.log(reason);
        done();
    });
});

test('Lang-server is started properly', () => {
    expect(langClient.initializedResult.capabilities.experimental.astProvider).toBe(true);
});

function testDiagramRendering(ast: BallerinaAST,  uri: string) {
    function getAST(url: string) {
        return Promise.resolve({ ast });
    }
    function parseFragment(fragment: any) {
    }
    function goToSource(pos: any) {
    }
    function onChange(ast: any) {
    }
    function getEndpoints() {
    }

    var getTextWidth = jest.fn();
    getTextWidth.mockReturnValue(100);
    jest
        .spyOn(SizingUtil.prototype, 'getTextWidth')
        .mockImplementation(getTextWidth);

    var consoleLog = jest.fn();
    jest
        .spyOn(console, 'log')
        .mockImplementation(consoleLog);

    const component = create(
        <BallerinaDiagramWrapper 
            docUri={uri}
            getAST={getAST}
            onChange={onChange}
            width={1000}
            height={1000}
            getEndpoints={getEndpoints}
            parseFragment={parseFragment}
            goToSource={goToSource}
        />
    );
    component.toJSON();
    expect(getTextWidth).toHaveBeenCalled();
}

var bbeFiles = globSync(path.join(bbeDir, '**', 'hello_world_client.bal'), {});
bbeFiles.forEach((file) => {
    const uri = URI.file(file).toString();
    test('Parsing and Rendering BBE: ' + path.basename(file), (done) => {
        langClient.getAST({
            documentIdentifier: {
                uri 
            }
        }).then((resp) => {
            expect(resp.parseSuccess).toBeTruthy();
            expect(resp.ast).toBeDefined();
            testDiagramRendering(resp.ast, uri);
            done();
        }, (reason) => {
            console.log('Error: ' + reason);
            done();
        });
    });
});

afterAll(() => {
    if (langClient) {
        langClient.kill();
    }
});