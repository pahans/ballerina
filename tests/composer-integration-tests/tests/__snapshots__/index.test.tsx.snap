// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parsing and Rendering BBE: hello_world_client.bal 1`] = `
<div
  className="ui active transition visible inverted dimmer"
  onClick={[Function]}
>
  <div
    className="content"
  >
    <div
      className="ui large loader"
    />
  </div>
</div>
`;

exports[`Parsing and Rendering BBE: hello_world_client.bal 2`] = `
<div
  className="ballerina-editor design-view-container"
>
  <div
    className="design-view-container "
    style={
      Object {
        "display": "block",
      }
    }
  >
    <div
      className="outerCanvasDiv"
    >
      <div />
      <div
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        style={
          Object {
            "height": 1000,
            "overflow": "hidden",
            "position": "relative",
            "width": 1000,
          }
        }
      >
        <div
          style={
            Object {
              "WebkitOverflowScrolling": "touch",
              "bottom": 0,
              "left": 0,
              "marginBottom": 0,
              "marginRight": 0,
              "overflow": "scroll",
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <div
            className="canvas-container"
          >
            <div
              className="canvas-top-controls-container"
            />
            <div
              className="html-overlay"
            />
            <div
              className="graphical-editor"
            />
            <div
              className="diagram root grabbable"
            >
              <div
                className="design-view-error"
              >
                <p>
                  Oops. Something went wrong.
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          className="scrollbar scrollbar-track-horizontal"
          style={
            Object {
              "display": "none",
              "height": 6,
              "position": "absolute",
            }
          }
        >
          <div
            className="scrollbar-thumb-horizontal"
            style={
              Object {
                "display": "block",
                "height": "100%",
                "position": "relative",
              }
            }
          />
        </div>
        <div
          className="scrollbar scrollbar-track-vertical"
          style={
            Object {
              "display": "none",
              "position": "absolute",
              "width": 6,
            }
          }
        >
          <div
            className="scrollbar-thumb-vertical"
            style={
              Object {
                "display": "block",
                "position": "relative",
                "width": "100%",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
